Add support for NSURLCache.

kgoodier opened the pull request:
The NSURL loading system has subtle and poorly documented behavior
regarding caching. CocoaSPDY was not doing the right thing, thus
there was no support for NSURLCache in the protocol. This patch
adds basic support for both NSURLConnection and NSURLSession based
requests. This is not yet a fully-featured client caching
implementation.

If the request specifies a NSURLRequest cachePolicy of:

NSURLRequestUseProtocolCachePolicy
- NSURL system does not provide a cached response to the protocol
  constructor. It is up to the protocol to load and validate the
  cached response.

NSURLRequestReturnCacheDataElseLoad
- NSURL system will provide the cached response, if available. The
  protocol is expected to validate the response, and load if not
  available or not valid.

NSURLRequestReturnCacheDataDontLoad
- NSURL system will provide the cached response, if available. The
  protocol is expected to validate the response. The protocol will not
  be loaded if no cached response is available, or if one is but is
  invalid, the protocol should not load the request.

In the cases where the protocol has a cached response and it is valid,
it is supposed to call URLProtocol:cachedResponseIsValid. CocoaSPDY
was not doing this. Determining validity of the cached response is the
job of the protocol, and a basic implementation has been provided here.

CocoaSPDY also has to jump through some hoops whenever NSURLSession is
being used, as Apple has not provided a way to get the
NSURLSessionConfiguration and thus we cannot get the right NSURLCache.
Fortunately we have already provided a workaround for this.

SPDYMetadata has been extended to provide the source of the response.

Observationally, we've seen the NSURL loading system fail to insert
items into the NSURLCache when using NSURLSession and iOS 7/8.
NSURLConnection works fine on all these, and NSURLSession works fine on
iOS 9. Best guess is there's a bug in the NSURLProtocol implementation
that fails to insert into the cache.

So, here we are creating a workaround for that specific case. CocoaSPDY
will buffer the data internally, up to a certain size limit based on the
cache size, and will manually insert into the cache when the response is
done. This could potentially be expanded in the future for unclaimed,
finished push responses, but for now those are excluded.

kgoodier committed:
Add support for NSURLCache.

The NSURL loading system has subtle and poorly documented behavior
regarding caching. CocoaSPDY was not doing the right thing, thus
there was no support for NSURLCache in the protocol. This patch
adds basic support for both NSURLConnection and NSURLSession based
requests. This is not yet a fully-featured client caching
implementation.

If the request specifies a NSURLRequest cachePolicy of:

NSURLRequestUseProtocolCachePolicy
- NSURL system does not provide a cached response to the protocol
constructor. It is up to the protocol to load and validate the
cached response.

NSURLRequestReturnCacheDataElseLoad
- NSURL system will provide the cached response, if available. The
protocol is expected to validate the response, and load if not
available or not valid.

NSURLRequestReturnCacheDataDontLoad
- NSURL system will provide the cached response, if available. The
protocol is expected to validate the response. The protocol will not
be loaded if no cached response is available, or if one is but is
invalid, the protocol should not load the request.

In the cases where the protocol has a cached response and it is valid,
it is supposed to call URLProtocol:cachedResponseIsValid. CocoaSPDY
was not doing this. Determining validity of the cached response is the
job of the protocol, and a basic implementation has been provided here.

CocoaSPDY also has to jump through some hoops whenever NSURLSession is
being used, as Apple has not provided a way to get the
NSURLSessionConfiguration and thus we cannot get the right NSURLCache.
Fortunately we have already provided a workaround for this.

SPDYMetadata has been extended to provide the source of the response.

kgoodier committed:
Workaround for caching with iOS 8 and NSURLSession.

Observationally, we've seen the NSURL loading system fail to insert
items into the NSURLCache when using NSURLSession and iOS 7/8.
NSURLConnection works fine on all these, and NSURLSession works fine on
iOS 9. Best guess is there's a bug in the NSURLProtocol implementation
that fails to insert into the cache.

So, here we are creating a workaround for that specific case. CocoaSPDY
will buffer the data internally, up to a certain size limit based on the
cache size, and will manually insert into the cache when the response is
done. This could potentially be expanded in the future for unclaimed,
finished push responses, but for now those are excluded.

This change also adds a few additional caching-related unit tests.

kgoodier committed:
Fix caching on iOS 7

On iOS 7, the protocol is required to *always* load its own cache item,
even for the NSURLRequestReturnCacheDataDontLoad and
NSURLRequestReturnCacheDataElseLoad policies. This applies to both
NSURLConnection and NSURLSession.

Found by unit testing on iOS 7 simulator with XCode 6.4.

NSProgrammer commented:
Looks great!  I have a number of comments that should be addressed though.  Thanks, Kevin!

kgoodier committed:
Caching code cleanup per code review feedback.

Simplify some things, handle edge cases, handle OSX.

NSProgrammer commented:
+1

niw commented on SPDYUnitTests/SPDYMockSessionManager.m:
```diff
@@ -0,0 +1,83 @@
+    if (self != nil) {
+    }
+    return self;
+}
```
Do we need this init override?

kgoodier replied:
Nope. :) Leftover.

niw commented on SPDY/SPDYCacheStoragePolicy.m:
```diff
@@ -13,6 +13,59 @@
+
+        for (int format = 0; (size_t)format < (sizeof(kTimeFormatInfos) / sizeof(kTimeFormatInfos[0])); format++) {
+            HTTPTimeFormatInfo info = kTimeFormatInfos[format];
+            if (info.readFormat != NULL && strptime(utf8String, info.readFormat, &parsedTime)) {
```
`strptime(3)` doesn't initialize `struct tm`, should be clear before using it, or, some of value may be undefined.
`strptime(3)` using current system locale, which is unknown. must be `strptime_l(3)` ?

kgoodier replied:
Done.

niw commented on SPDY/SPDYCacheStoragePolicy.m:
```diff
@@ -13,6 +13,59 @@
+    for (NSString *component in components) {
+        NSArray *pair = [component componentsSeparatedByString:@"="];
+        NSString *key = [pair[0] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
+        NSString *value = pair.count == 2 ? [pair[1] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] : @"";
```
This is not right parsing. by RFC definition, `Cache-Control` is defined like

>    Cache-Control   = 1#cache-directive
>    cache-directive = token [ "=" ( token / quoted-string ) ]

and `quoted-string` can contains `=`. We must like concatenating `pair[1]` to last one with `=`.

niw commented on SPDY/SPDYCacheStoragePolicy.m:
```diff
@@ -83,3 +150,71 @@ extern NSURLCacheStoragePolicy SPDYCacheStoragePolicy(NSURLRequest *request, NSH
+    // ...
+
+    return SPDYCachedResponseStateValid;
+}
```
Add a new line?

kgoodier committed:
Better parsing of Cache-Control header.

Feedback from code review:
- parse quoted strings
- use strptime_l
- zero tm struct for strptime_l
- allow requests with Authorization to be cached

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/CocoaSPDY?pullRequest=141) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/twitter/CocoaSPDY?pullRequest=141) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/CocoaSPDY?pullRequest=141) it.</sub>
